wgengine/netstack/netstack.go

```diff
@@ -381,18 +382,81 @@ func (ns *Impl) acceptUDP(r *udp.ForwarderRequest) {
 		ns.logf("Could not create endpoint, exiting")
 		return
 	}
+	localAddr, err := ep.GetLocalAddress()
+	if err != nil {
+		return
+	}
+	remoteAddr, err := ep.GetRemoteAddress()
+	if err != nil {
+		return
+	}
 	c := gonet.NewUDPConn(ns.ipstack, &wq, ep)
-	go echoUDP(c)
+	go ns.forwardUDP(c, &wq, localAddr, remoteAddr)
 }
 
-func echoUDP(c *gonet.UDPConn) {
-	buf := make([]byte, 1500)
-	for {
-		n, err := c.Read(buf)
+func (ns *Impl) forwardUDP(client *gonet.UDPConn, wq *waiter.Queue, clientLocalAddr, clientRemoteAddr tcpip.FullAddress) {
+	port := clientLocalAddr.Port
+	ns.logf("[v2] netstack: forwarding incoming UDP connection on port %v", port)
+	backendLocalAddr := &net.UDPAddr{Port: int(clientRemoteAddr.Port)}
+	backendRemoteAddr := &net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: int(port)}
+	backendConn, err := net.ListenUDP("udp4", backendLocalAddr)
+	if err != nil {
+		ns.logf("netstack: could not bind local port %v: %v, trying again with random port", clientRemoteAddr.Port, err)
+		backendConn, err = net.ListenUDP("udp4", nil)
 		if err != nil {
-			break
+			ns.logf("netstack: could not connect to local UDP server on port %v: %v", port, err)
+			return
+		}
+	}
+	var cancels [2]context.CancelFunc
+	timer := time.AfterFunc(2*time.Minute, func() {
+		ns.logf("netstack: forwarder UDP connection on port %v closed", port)
+		for _, c := range cancels {
```

Comment:
I think this is technically racy. You read these in one goroutine and write them in another with no synchronization other than time (which isn't synchornization)

But this is also all unnecessary I think.

Instead of the [2]CancelFunc, just have one ctx you pass down to both copies.

Do a `ctx, cancel := context.WithCancel(context.Background())` before time.AfterFunc and then your AfterFunc func can just `cancel()`.